cache_toolbox/core.py

```diff
@@ -42,53 +102,75 @@ def get_instance(model, instance_or_pk, timeout=None, using=None):
     """
 
     pk = getattr(instance_or_pk, 'pk', instance_or_pk)
-    key = instance_key(model, instance_or_pk)
-    data = cache.get(key)
 
-    if data is not None:
+    primary_model = model
+    descriptors = getattr(primary_model, '_cache_fetch_related', ())
+    models = [model, *(d.related.field.model for d in descriptors)]
+    # Note: we're assuming that the relations are primary key foreign keys, and
+    # so all have the same primary key. This matches the assumption which
+    # `cache_relation` makes.
+    keys_to_models = {
+        instance_key(model, instance_or_pk): model
+        for model in models
+    }
+
+    data_map = cache.get_many(keys_to_models.keys())
+    instance_map = {}
+
+    if data_map:
         try:
-            # Try and construct instance from dictionary
-            instance = model(pk=pk, **pickle.loads(data))
-
-            # Ensure instance knows that it already exists in the database,
-            # otherwise we will fail any uniqueness checks when saving the
-            # instance.
-            instance._state.adding = False
-
-            # Specify database so that instance is setup correctly. We don't
-            # namespace cached objects by their origin database, however.
-            instance._state.db = using or DEFAULT_DB_ALIAS
-
-            return instance
+            for key, data in data_map.items():
+                model = keys_to_models[key]
+                instance_map[key] = deserialise(model, data, pk, using)
         except:
             # Error when deserialising - remove from the cache; we will
             # fallback and return the underlying instance
-            cache.delete(key)
+            cache.delete_many(keys_to_models.keys())
```

Comment:
This is fairly blunt; should we attempt only to delete the ones which we actually know to be broken?